### PR TITLE
Expose problem definition and context separation on Tooling API events

### PR DESCRIPTION
**Pull request by donat**

### CODE REVIEW

**Comment by reinsch82** [comment]
@bot-gradle test this

**Comment by reinsch82** [comment]
@bot-gradle test ACT

**Comment by reinsch82** [comment]
@bot-gradle test RFN

File: platforms/ide/tooling-api/src/main/java/org/gradle/tooling/internal/protocol/problem/InternalProblemDetailsVersion2.java [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.tooling.internal.protocol.problem;
+
+public interface InternalProblemDetailsVersion2 {
+
```
Comment (reinsch82) [inline platforms/ide/tooling-api/src/main/java/org/gradle/tooling/internal/protocol/problem/InternalProblemDetailsVersion2.java:20]: @donat I removed the second one in `org.gradle.tooling.internal.protocol`, is that ok?

**Comment by reinsch82** [comment]
@bot-gradle test this

**Comment by reinsch82** [comment]
@bot-gradle test this

File: platforms/ide/tooling-api-builders/src/main/java/org/gradle/tooling/internal/provider/runner/AggregatingProblemConsumer.java [hunk 2]
```diff
@@ -57,7 +59,7 @@ void setThresholdForIntermediateSummary(int thresholdForIntermediateSummary) {
     }
 
     void sendProblemSummaries() {
-        List<InternalProblemAggregationVersion2> problemSummaries = createSummaries();
+        List<InternalProblemAggregationVersion3> problemSummaries = createSummaries();
```
Comment (donat) [inline platforms/ide/tooling-api-builders/src/main/java/org/gradle/tooling/internal/provider/runner/AggregatingProblemConsumer.java:62]: @reinsch82 I remember I didn't thoroughly checked that my changes regarding aggregation works as you intended. Can you please verify that this PR preserves the problem aggregation feature?

File: platforms/ide/tooling-api/src/main/java/org/gradle/tooling/events/problems/SingleProblemEvent.java [hunk 3]
```diff
@@ -28,6 +28,15 @@
...
 public interface SingleProblemEvent extends ProblemEvent {
+
+    /**
+     * Returns the problem definition.
+     *
+     * @return the definition
+     * @since 8.8
+     */
+    ProblemDefinition getDefinition();
+
```
Comment (donat) [inline platforms/ide/tooling-api/src/main/java/org/gradle/tooling/events/problems/SingleProblemEvent.java:39]: @reinsch82 wdyt, should we remove outdated categories like `#getCategory()`? It's a potential breaking change. At the same time, this is an incubating API?

Comment (reinsch82) [reply platforms/ide/tooling-api/src/main/java/org/gradle/tooling/events/problems/SingleProblemEvent.java:39]: I would support removing outdated APIs, we never get rid of it once it's official

File: platforms/ide/tooling-api/src/main/java/org/gradle/tooling/internal/protocol/problem/InternalProblemDetailsVersion2.java (hunk 1 above)
Comment (donat) [reply platforms/ide/tooling-api/src/main/java/org/gradle/tooling/internal/protocol/problem/InternalProblemDetailsVersion2.java]: That make sense.

**Comment by donat** [comment]
@bot-gradle test ReadyForNightly

**Comment by donat** [comment]
@bot-gradle test this

**Comment by donat** [comment]
@bot-gradle test this

**Comment by donat** [comment]
@bot-gradle test this

**Comment by donat** [comment]
@bot-gradle test RDM

**Comment by donat** [comment]
@bot-gradle test RFM

**Comment by donat** [comment]
@bot-gradle test RFM

**Comment by donat** [comment]
@bot-gradle test RFN

**Comment by donat** [comment]
@bot-gradle merge

File: platforms/ide/tooling-api/src/main/java/org/gradle/tooling/events/problems/SingleProblemEvent.java (hunk 3 above)
Comment (donat) [reply platforms/ide/tooling-api/src/main/java/org/gradle/tooling/events/problems/SingleProblemEvent.java:39]: This PR is getting way too big. I'll merge it and do the cleanup separately.

**Pull request merged**